feat(auth): Implement Passwordless Authentication

raisedadead opened the pull request:
Feature Checklist:
- [x] Create API for creating/saving a temporary token
- [x] Add email template and sending logic
- [x] Add validations and timeouts
- [x] Update the sign in view
- [x] Prevent users without verified email from login
- [x] ~~Create API for creating/saving an access token~~ Using loopbacks defaults
- [x] Add routes for requesting access token
- [x] Invalidate used/expired tokens
- [x] Remove password reset logic
- [x] Handle all errors gracefully
- [x] Remove all passwords from database (#14559 issue created for DevOps)
- [x] Add debounce/disable to magic link button
- [x] Add some loading indicator to magic link button ajax request
- [ ] (berks): Investigate memory consumption issue

Closes #8165 
Closes #9027
Closes #15254

raisedadead committed:
Implement passwordless login

* Created a new rest API to create and save a temporary token that
  can be exchanged for a access token.
* Updated the sign in view
* Add email template for sign in links
* Add route to request a access token and login user
* Make the email views conistent

QuincyLarson commented:
@atjonathan @raisedadead @BerkeleyTrue I have created a new issue related to this here: https://github.com/FreeCodeCamp/FreeCodeCamp/issues/10407

BerkeleyTrue commented:
@raisedadead Please continue to work on this. #10407 does not change the user stories of this PR.

QuincyLarson commented:
Reopening this because the underlying passwordless functionality is still needed for updating the workflow.

QuincyLarson commented:
@raisedadead I know you're recovering from the flu. I just wanted to check and see how this is coming along, and let you know that going passwordless is still a priority for us 😄

raisedadead commented:
Hi @quincylarson, just got back. I plan to work on this from tomorrow. Sorry about the delay.

QuincyLarson commented:
@raisedadead no need to apologize - thanks for helping us implement this!

raisedadead commented:
Just a rebase of the mess that I created. Will be ready with the feature in the next push.

BerkeleyTrue commented:
👍

raisedadead commented:
Basic functionality is there out now… just some more validation touches and code optimisations and we are good (soon).

QuincyLarson commented:
@raisedadead awesome - thank you for keeping us posted 😄

raisedadead committed:
Clean up remote pre-hooks because they wont be called.

raisedadead committed:
Optimize code and streamline flow.

- Optimize code and streamline flow.
- Add return statements for error display.
- Invalidate status vars and TTLs on User Model.
- Build Observable for findOrCreate

raisedadead committed:
Add invalidate token handler

raisedadead commented:
This is mostly ready, but am at my wits end for getting the request to complete when the link in the email is completed.

Basically I need to get the middleware `invalidateAuthToken` function to execute correctly.

Would appriciate a review and guidance from anyone.

raisedadead commented:
Never mind my last comment. I figured out that I was just more sleepy than I should have been last night.

This is ready for QA.

BerkeleyTrue commented on common/models/user.js:
```diff
@@ -382,14 +396,114 @@ module.exports = function(User) {
+
+          const { id: loginToken } = token;
+          const loginEmail = user.email;
+          const renderAuthEmail = loopback.template(path.join(
```
Instead of creating this renderer per request, can we move this up into the module scope and then use the ternary above to select the appropriate renderer instead of just the template name?

BerkeleyTrue commented on common/models/user.js:
```diff
@@ -382,14 +396,114 @@ module.exports = function(User) {
+          };
+          if (!isDev) {
+            this.email.send(mailOptions, err =>{
+              if (err) { throw err; }
```
Async errors must be handled correctly, throwing here will cause servers to crash unexpectedly.

BerkeleyTrue commented on common/models/user.js:
```diff
@@ -382,14 +396,114 @@ module.exports = function(User) {
+            console.log('~~~~\n' + mailOptions.text + '~~~~\n');
+          }
+          user.emailAuthLinkTTL = token.created;
+          user.save(err =>{ if (err) { throw err; }});
```
You must use the update$ methods when modifying the user object.

BerkeleyTrue commented on common/models/user.js:
```diff
@@ -382,14 +396,114 @@ module.exports = function(User) {
+
+        return dedent`
+           If you entered a valid email, a magic link is on its way.
+           Please follow that link to sign in.`;
```
indenting

BerkeleyTrue commented on common/models/user.js:
```diff
@@ -382,14 +396,114 @@ module.exports = function(User) {
+      .map((msg) => {
+        if (msg) { return msg; }
+        return dedent`
+         Oops, something is not right, please try again later.`;
```
indenting

BerkeleyTrue commented on common/models/user.js:
```diff
@@ -382,14 +396,114 @@ module.exports = function(User) {
+        return dedent`
+         Oops, something is not right, please try again later.`;
+      })
+      .catch(error => {
```
Errors must be dealt with by express here. Do not use a catch here unless modifying the error message and then rethrowing

raisedadead replied:
I agree, will look into an alternate way for this. If you remember we did this for returning the error to the Promise (AJAX) from the client.

BerkeleyTrue commented on common/models/user.js:
```diff
@@ -402,17 +516,17 @@ module.exports = function(User) {
-      return Observable.throw(new Error(
-        `Please wait ${timeToWait} to resend email verification.`
-      ));
+    if (!isDev) {
```
isDev shouldn't be used like this. dev and production must run the same.

BerkeleyTrue commented on server/boot/user.js:
```diff
@@ -152,6 +153,9 @@ module.exports = function(app) {
     );
   }
 
+  AccessToken.findOne$ = Observable.fromNodeCallback(
```
indenting

raisedadead committed:
Fix indenting and remove isDev from wait time.

raisedadead committed:
Donot render per request email

raisedadead committed:
Remove error catch and use update method

raisedadead committed:
Unify signin and signup with lesser clicks

raisedadead committed:
Remove sign up page and update button styles

raisedadead committed:
Remove reset-password logic

raisedadead committed:
Remove non-react Update email views and routes

BerkeleyTrue commented on common/models/user.js:
```diff
@@ -382,14 +371,105 @@ module.exports = function(User) {
+      emailVerified: false
+    };
+    return User.findOrCreate$({ where: { email }}, userObj)
+      .map(([ err, user, isCreated ]) => {
```
Errors here will be passed down the pipeline until a catch operator is hit or a the onError handler in the subscribe method is called. The err is not passed into the map operator.

raisedadead committed:
chore(passwordless): Update to freeCodeCamp

raisedadead committed:
Update email views and notifications

raisedadead committed:
refactor(code): cleanup and rebase

raisedadead committed:
fix(email): allow sender email var for development

Allows setting the SES_ID, SES_SECRET and EMAIL_SENDER values

raisedadead committed:
fix: Add error handling and refactoring of Observable methods

raisedadead committed:
fix(email): Error messages for updateEmail API

raisedadead commented:
@BerkeleyTrue @QuincyLarson This is ready for a QA. 
Thanks a ton to Berkeley for taking the time to help me through this.

I have discovered a few issues with the updateEmail API while doing a load test, and will fix that in a separate PR, because its unrelated and we have open issues for them I checked.

Nonetheless this is ready for a review and deployment on beta for a testing. Berkeley I could not reproduce the issue about bad tokens, maybe beta is a good candidate for a try out!

Also I have created a DevOps Issue for deleting the passwords, which Berkeley can take up when this reaches production.

Awaiting feedback! Cheers.

augmt commented:
@raisedadead In accordance with #10407, I think the sign-in/sign-up page should illustrate to campers that signing up for a freeCodeCamp account with GitHub is not allowed.

https://www.freecodecamp.com/signin currently does this best as it does not force returning campers to click an extra time like https://beta.freecodecamp.com/en/signup does and it separates supported sign-in methods from supported sign-up methods. What do you think?

raisedadead commented:
@augmt That is correct, and I plan to make a separate PR as soon as we complete testing this one. We are limiting the change in this PR to email auth and updates.

BerkeleyTrue commented:
At the moment, links sent out by beta in the email do not add the subdomain `beta`. This is not the fault of this PR but we will need to add some sort of feature to include this ability before we can test this on Beta. We can either do it here or as a separate PR.

raisedadead commented:
@BerkeleyTrue I'll add it in this PR itself as soon as I am back tonight. Thanks for the review.

raisedadead committed:
feat(server): Add URL utils to compute server URL, HOST, PORT, etc.

raisedadead commented:
Hey @BerkeleyTrue I have created a new utility for future use to compute the URLs including things like HOST, PORT etc.

Its a overkill as of now, but I am sure would help in the long run.

Finally I have made a minor change in a previous commit to this PR ensuring that at least the message for a success case reaches the client while updating email.

Please check and let me know how this looks.

BerkeleyTrue commented on server/utils/url-utils.js:
```diff
@@ -0,0 +1,29 @@
+}
+
+export function getPort() {
+  return isDev ? (process.env.SYNC_PORT || 3000) : 443;
```
nit: use a if/esle condition for readability here. 
fix: if not dev port should be omitted.

BerkeleyTrue commented on server/utils/url-utils.js:
```diff
@@ -0,0 +1,29 @@
+        + '://'
+        + getHost()
+        + ':'
+        + getPort();
```
fix: If not dev this should not include port number or the column.

raisedadead committed:
fix(email): Add a trailing slash to avoid redirects

raisedadead commented:
Done!

BerkeleyTrue commented on server/utils/url-utils.js:
```diff
@@ -0,0 +1,37 @@
+
+export function getPort() {
+  if (!isDev) {
+    return '443';
```
fix: should not add port in production. This should be null

raisedadead replied:
Ah, cool.. I'll update, I used this from the previous code for the verify email send logic, where I think the port was used for production. But NVM, I'll update asap.

BerkeleyTrue replied:
ping

BerkeleyTrue commented on server/utils/url-utils.js:
```diff
@@ -0,0 +1,37 @@
+       + '://'
+       + getHost()
+       + ':'
+       + getPort();
```
fix: This should not include port in production

raisedadead replied:
Yup, isn't the production case in the conditional, i.e the previous block?

BerkeleyTrue replied:
Ah, I see. I was pattern matching with the previous function.

BerkeleyTrue commented on server/utils/url-utils.js:
```diff
@@ -0,0 +1,37 @@
+}
+
+export function getServerFullURL() {
+  if (!isDev) {
```
Let's swap these two conditionals so they read a little more obviously.

if (isDev) {
}

BerkeleyTrue commented:
Through this up on beta and there are still a couple of issues: 

* link sent in email is still pointing to freecodecamp and not beta
* If I press the 'get magic link` button several times in a row, it will send me several emails back to back
     * (client side) to prevent this we need to disable the button while there is a request in flight
     * (server side) We need to figure out some way to lock the user object to prevent the user from constantly sending out emails (unchecked this could provide a vulnerability that could crash servers)

raisedadead commented:
> link sent in email is still pointing to freecodecamp and not beta

I tested by setting the below, could NOT reproduce:

```
NODE_ENV=production
BETA=true
```
The link is correctly generated, for me.

> If I press the 'get magic link` button several times in a row, it will send me several emails back to back

Yup, we need a strong mechanism for this. 

Additionally, I am seeing a completely different issue, where createAccessToken from loopback is not correctly associating the correct userId to the accessToken that it generates.

This is a weird one.

Can you hit me up in chat when you can?

raisedadead commented:
@BerkeleyTrue 

Regarding the second concern for multiple request, I have two proposals:

1. Disable the submit, button while the AJAX is in progress.
2. Migrate the Sign up page to react and use RxJS debounce.

I'll let you suggest which one would be a better alternative?

BerkeleyTrue commented:
The problem lies mainly server side. The question is how do you lock a req path by the same user? Like transactions with databases only for expressjs routes.

BerkeleyTrue commented:
Maybe we can implement some sort of queuing system to send the emails?

BerkeleyTrue commented:
Looks like the beta flag issue is solved.

raisedadead committed:
fix(email-auth): add disable and loading indication to button

raisedadead commented:
@BerkeleyTrue: 
- Added the disable and loading indication to the button when the AJAX is in flight. 
- For additional debouncing, I think we can take that at the server side rather.

BerkeleyTrue commented:
@raisedadead I think we can merge this as soon as the merge conflicts are done. I'll run through it one more time and then we will through it up on beta

raisedadead commented:
This is just an initial rebase of the conflicts, which I basically eye-balled. I am yet to check the changes that I would need to make to get this working as it was earlier.  :rotating_light: :hand:

raisedadead committed:
fix(auth): Fix merge errors from rebase

BerkeleyTrue commented:
@raisedadead Somehow all the refactoring I did in user.js has been undone in this PR

raisedadead commented:
@BerkeleyTrue yes, that is correct. I might have messed up the rebase a bit. And have been away a while on an outing. Please give a day or so to clean this up correctly. I'll let you know as soon as it ready.

BerkeleyTrue commented:
@raisedadead ping

frutsy007 commented:
How is work

BerkeleyTrue commented:
@piotrek1543 @frutsy007 refrain from unrequested PR review approvals and unrelated comments.

raisedadead commented:
Hey @BerkeleyTrue so sorry about the delay, I have been a swamped a bit lately for the past couple weeks. I plan to work on this tonight.

BerkeleyTrue commented:
@raisedadead no worries. Thanks for the update. I'm excited to see this merged in!

BerkeleyTrue commented:
woot!!

raisedadead commented:
@BerkeleyTrue this is ready for QA. Please check when you can. Thanks.

BerkeleyTrue commented:
Looks good. I'll merge when I get to a computer

BerkeleyTrue commented:
merged!!!!!!!!!!!!!!!!!!!!
https://github.com/freeCodeCamp/freeCodeCamp/commit/61d740b8a0fb144e77a6fa963d0ec7fe16cfead4

✨ 🎊 🚀

I owe you a 🍺 

Happy Coding!

raisedadead commented:
@BerkeleyTrue awesome, thanks a lot for the QA. Do let me know how it goes when deployed? 
Of course, let's grab some 🍺 when we do meet in person in future!  😄
